src/libply/libply.c

```diff
@@ -360,6 +377,63 @@ static int ply_load_bpf(struct ply *ply)
 	return err;
 }
 
+/* Used to verify that bpf_prog_test_run can work correctly to detect
+   whether the current environment supports BEGIN/END.  */
+static int ply_check_prog_test_run()
+{
+	static int has_result = 0;
+	static int can_prog_test_run_ = 0;
+
+	if (has_result)
+		return can_prog_test_run_;
+
+	int fd;
+	size_t vlog_sz = 0;
+	char *vlog = NULL;
+
+	struct bpf_insn *insns = NULL;
+
+	vlog = malloc(1 << 12);
+	insns = malloc(2 * sizeof(*insns));
+
+	if (vlog && insns)
+		vlog_sz = 1 << 12;
+	else
+		_w("not enough memory to enable the kernel verifier output\n");
+
+	insns[0] = INSN(BPF_ALU64 | BPF_MOV | BPF_K, BPF_REG_0, 0, 0, 0);
+	insns[1] = EXIT;
+
+	fd = bpf_prog_load(BPF_PROG_TYPE_RAW_TRACEPOINT,
+			   insns, 2, vlog, vlog_sz);
+
+	free(insns);
+	if (fd < 0) {
+		_e("unable to load raw tracepoint, errno:%d\n", errno);
+		if ((errno == EINVAL) && vlog && !vlog[0])
+			_w("was ply built against the running kernel?\n");
+		else if (vlog && vlog[0])
+			_e("output from kernel bpf verifier:\n%s\n", vlog);
+		else
+			_e("no output from kernel bpf verifier, "
+			   "retry with debugging enabled\n");
+	}
+
+	if (vlog)
+		free(vlog);
+
+	if (!bpf_prog_test_run(fd))
+		can_prog_test_run_ = 1;
+
+	close(fd);
+	has_result = 1;
+
+	if (!can_prog_test_run_)
+		_w("The current environment does not support BEGIN/END\n");
+
+	return can_prog_test_run_;
+}
+
```

Comment:
I don't think we need this. We already know that `BPF_PROG_TEST_RUN` is defined, so this should all work. If something isn't working, `ply -T` should already be able to pinpoint the issue.

Reply:
Indeed, fixed in [93bbc1b](https://github.com/iovisor/ply/pull/82/commits/93bbc1b022b4c2587e0b32aa3fd52bb752da5e2a) .
Please take a look, thanks.
Tested after modification, the result is normal:
```
$ sudo ply -T
Verifying kernel config (/boot/config-5.15.0-53-generic)... OK
Ensuring that debugfs is mounted... OK
Verifying kprobe... OK
Verifying tracepoint... OK
Verifying special... OK
Verifying interval... OK
```